### PR TITLE
Refactored classical_gradients test to pytest

### PR DESCRIPTION
**Pull request by quantshah**
**Description of the Change:**

This refactors and gets rids of the loops for the classical gradient tests. 

**Benefits:**
Tests are self-contained, easier to read

**Possible Drawbacks:**
Repetition of some code but they could be parameterized with pytest fixtures.

**Related GitHub Issues:**
#99

### CODE REVIEW

**Comment by quantshah** [comment]
Why are we testing with three different functions for each case? A proper unit test should test one thing and fail for one case right? I understand that this can be done with the pytest fixture to make it compact, but still why do we have to make three functions everytime to test this instead of testing with just one.

**Comment by josh146** [comment]
> Why are we testing with three different functions for each case? A proper unit test should test one thing and fail for one case right? I understand that this can be done with the pytest fixture to make it compact, but still why do we have to make three functions everytime to test this instead of testing with just one.

This is correct, we shouldn't be doing this. Each function should be a separate unit test. Refactoring to fix this can be considered part of #99

**Comment by josh146** [comment]
I can't quite work this out, but for some reason the Python 3.6 test fails (even though 3.5 and 3.7 pass), with the following failure:
```
=================================== FAILURES ===================================
_______________________ test_gradient_multivar_multidim ________________________
Traceback (most recent call last):
  File "/home/travis/build/XanaduAI/pennylane/tests/test_classical_gradients.py", line 183, in test_gradient_multivar_multidim
    np.allclose(auto_grad, correct_grad)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/autograd/tracer.py", line 48, in f_wrapped
    return f_raw(*args, **kwargs)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/numpy/core/numeric.py", line 2423, in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/numpy/core/numeric.py", line 2522, in isclose
    yfin = isfinite(y)
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

**Comment by co9olguy** [comment]
@josh146 seems to be a numpy issue. Is the numpy version the same for each version of python?

**Comment by co9olguy** [comment]
What is the actual type of the objects that lead to the error? i.e., `auto_grad` and `manual_grad`?

**Comment by josh146** [comment]
> @josh146 seems to be a numpy issue. Is the numpy version the same for each version of python?

This could be the issue. Travis caches the pip environment, and could lead to different dependency versions in the different jobs

File: tests/test_classical_gradients.py [hunk 1]
```diff
@@ -15,129 +15,196 @@
...
-if __name__ == '__main__':
-    print('Testing PennyLane version ' + qml.version() + ', classical gradients.')
-    # run the tests in this file
-    suite = unittest.TestSuite()
-    for t in (BasicTest,):
-        ttt = unittest.TestLoader().loadTestsFromTestCase(t)
-        suite.addTests(ttt)
-
-    unittest.TextTestRunner().run(suite)
+
```
Comment (josh146) [inline tests/test_classical_gradients.py]: Lets set a seed at the top of the file (`np.random.seed(42)`) to force the tests to be deterministic.

File: tests/test_classical_gradients.py [hunk 2]
```diff
@@ -15,129 +15,196 @@
...
+def test_gradient_univar(tol):
```
Comment (josh146) [inline tests/test_classical_gradients.py]: To address your comment on multiple tests per test function, what you could do here is make each of the existing test functions a _class_, and then break the contents of each class into test methods, e.g.,

```python
class TestGradientUnivar:
    """Tests gradients of univariate unidimensional functions calculated via qml.grad."""

    def test_sin(self, tol):
        """tests that the gradient of the sin function is calculated correctly"""
        x_vals = np.linspace(-10, 10, 16, endpoint=False)
        g = qml.grad(np.sin, 0)
        auto_grad = [g(x) for x in x_vals]
        correct_grad = np.cos(x_vals)
        assert np.allclose(auto_grad, correct_grad, atol=tol, rtol=0)
```

Comment (quantshah) [reply tests/test_classical_gradients.py]: Thanks for the suggestion. The tests look much better now. This is ready for review.

**Pull request merged**